### PR TITLE
improve state db flow

### PR DESCRIPTION
**Pull request by NoelBright**
move stateGenerator.go to db folder. Use GenerateStateRoot for mining.go
and store.go

Signed-off-by: NoelBright <noel.n.bright@gmail.com>

### Type (put an `x` where ever applicable)
- [ ] Bug fix: Link to the issue
- [x] Feature (Non-breaking change)
- [ ] Feature (Breaking change)
- [ ] Documentation Improvement

### CODE REVIEW

File: chain/db/statedb.go [hunk 1]
```diff
@@ -263,3 +271,37 @@ func (sdb *StateDB) SetID(addr common.Uint160, id []byte) error {
...
+		}
+		acc.SetBalance(amount - value)
+	default:
+		return errors.New("UpdateBalance: invalid operation")
+	}
+
+	return nil
+}
+
+func (sdb *StateDB) IncrNonce(addr common.Uint160) error {
```
Comment (yilunzhang) [inline chain/db/statedb.go]: Can we guarantee that IncrNonce will not be called concurrently?

File: chain/db/store.go [hunk 2]
```diff
@@ -78,13 +78,23 @@ func (cs *ChainStore) InitLedgerStoreWithGenesisBlock(genesisBlock *block.Block)
...
 
-		root := cs.GetCurrentBlockStateRoot()
-		fmt.Println("---------", root)
-		cs.States, _ = NewStateDB(root, NewTrieStore(cs.GetDatabase()))
+		root, err := cs.GetCurrentBlockStateRoot()
+		if err != nil {
+			return 0, nil
+		}
+
+		fmt.Println("state root:", root.ToHexString())
```
Comment (yilunzhang) [inline chain/db/store.go]: Why not using log here?

File: chain/db/statedb.go (hunk 1 above)
Comment (NoelBright) [reply chain/db/statedb.go]: account is proccessed sequentially now.

File: chain/db/store.go (hunk 2 above)
Comment (NoelBright) [reply chain/db/store.go]: done

File: chain/db/store.go [hunk 3]
```diff
@@ -464,7 +372,10 @@ func (cs *ChainStore) persist(b *block.Block) error {
 		return err
 	}
 
-	cs.States = states
+	cs.States, err = NewStateDB(root, NewTrieStore(cs.GetDatabase()))
```
Comment (yilunzhang) [inline chain/db/store.go]: If we get error here, the above persist functions are already executed, but cs.States will be empty. Will that cause a problem?

Comment (NoelBright) [reply chain/db/store.go]: hm, how about panic when returning error?

**Pull request merged**